### ISSUE TITLE
The NUnit 3 breaks integration with the TeamCity

### ISSUE THREAD

**Issue by NikolayPianikov** [nunit/nunit#976]
As you probably know, TeamCity has pretty decent integration with NUnit, and this year we released a version which integrates with NUnit3 beta. Unfortunately, after recent release of NUnit3 beta 5 our integration is no longer working. While this is understandable risk, when you're providing an integration with beta version of some tool, still we think that some problems could be easily removed. Also beta in the name usually means the tool will not change much, as it is going to be released really soon. Sadly this was not the case here. So, speaking about our problems:

1) First and foremost - NUnit still does not have proper ways to detect its version. We propose to add -version argument or some similar approach. We're aware that nunit-console.exe started without arguments will output version, but what if you change this in the future versions? Is it reliable? Clear approach to versioning will allow vendors, who integrate with the tool, at least to fail gracefully providing user friendly error messages.

2) NUnit 3 beta 5 removed filtering of tests by categories. Corresponding command line arguments were replaced with -where. Given that previous versions of NUnit3 supported these arguments, and we already released a version which works with these versions, we had to workaround this problem with proper using of -where. This sounds like not a big deal, but -where argument has not been supported by previous versions, so now we have to detect the version of the tool, see point 1. 

3) Parallel tests are now on by default. For us, it looks like a dangerous approach, which potentially can cause unnecessary problems to users. Users who did not have any problems with performance can face instability problems with tests after upgrade to new NUnit. Moreover, our users (those who run tests on TeamCity build agents) often run tests on virtual machines having 1 core, and for such users it does not make much sense to enable parallelization by default. So they will not benefit from parallelization, but chances are high that they will suffer from instability problems. Currently we have to workaround this problem too, by providing --workers=1 argument by default.

4) nunit-console.exe is now called nunit3-console.exe. Well for us, those who look at the project from outside, this change does not look like justified. Sure there are some incompatibilities with command line and people will have problems with their scripts where they call nunit-console.exe directly. But, some incompatibilities, like removed filtering of tests, could be easily fixed. Others, like parallelization by default, also have issues, and could be re-thought. Also such a change raises questions about future versions. Are we going to see nunit4-console.exe soon? What about minor versions, do you plan to have nunit31-console.exe? Is this now official approach to versioning? Does it mean that people will have to fix their build scripts and CI server configuration, each time when you change executable name?

And as a general note, for vendors like us, command line arguments are like API, and as it makes sense to preserve compatibility with API, it also makes sense to preserve it with command line. Note, that we're all for new tools, new features, etc. But changes in API should be justified, and vendors and your users should have clear understanding why they need to fix their tools. And just to make our point more clear. It's not that we insist on backward compatibility, we understand that tool should evolve. It's just that sometimes, backward compatibility can be preserved with relatively small effort. And this small effort can save days or months of work of other people. It would be great if NUnit team paid a little bit more attention to compatibility issues in the future.

**Comment by oznetmaster** [comment]
I am responding solely for myself.

Using --workers=0 actually is a better solution since it disable the parallel mechanism entirely (everything functions as if PARALLEL had not been specified in the build).

NUnit 3.0 has been an evolutionary process.  Like any piece of software that is going through development with numerous alpha and beta releases, nothing is "guaranteed" about the software functionality, API, command interface, or output format, until it reaches release candidate status.

If the software is used prior to that time, the user or developer implicitly accepts the fact that each intermediate release will potentially require changes to their own software, or operation procedures, to match the changes in that intermediate release.

This is why many wait until the evolving software reaches release candidate status before adapting the new version of the software.

I do not agree that during this evolution process, the software developer of the evolving software should feel compelled to maintain any compatibility with a previous intermediate release, or to spend any resources trying to accommodate those who have been using those previous intermediate releases.  That would defeat the entire evolutionary development process.

To complain that some change is not "justified" seems rather self-serving.  During evolutionary development, all changes are justified.  That is the nature of evolution.  As long as those changes are well documented at each step of the evolution, the developer has met their full responsibility to those who are using the evolving product at these intermediate steps.

**Comment by pavelsher** [comment]
Hello,

I also work for JetBrains, so let me clarify this letter one more time. It's not like we complain, it's more like FYI and questions to the team at the same time. We took the risks to integrate with beta version and we understood at that point how this could affect us. 

But as a team you're making some decisions when you evolve the tool. Some of your decisions do not look justified for us, who integrate with it. We tried to describe recent problems so that you better understand what we have to do to integrate with new beta version. 

And I don't think it is selfish to tell you about these problems. Other vendors or your users can have the same issues. They probably already have scripts to start tests with different categories, all these scripts have to be changed if they decide to move to NUnit 3. Take a look at this from another angle - there are lots of people who use NUnit. And some incompatibility in a new version can easily translate to days if not months of work if we sum up all of the affected users. At the same time it looks like arguments for categories could've been left for better compatibility without any significant efforts from your team, because internally they should translate simply to new where filter. Please correct me if I am wrong here, I may not know all technical details. 

Thank you for the hint about workers, we'll apply it. 

It would be great to have better understanding on what is going on with versioning. Do you have clear vision how vendors like us should determine the version of NUnit in a reliable way?

**Comment by oznetmaster** [comment]
I will address the parallelization issue briefly.

As the person who does most of the support for the compact framework I pushed to have the CF build be PARALLEL (it had never been previously).  I got a lot of resistance because prior to CF 3.9, multi-core was not supported in CF.  So why allow parallel?

I noted that parallel is not only useful when there are multiple cores.  If the specific tests are heavily I/O bound, for instance, then allowing parallel tests to run can improve the testing performance significantly.

Since setting --workers=0 allows a fallback to the previous non-parallel environment, it seemed wasteful to have implemented all the mechanisms for parallel tests, and the control of that parallelization, and not allow it for all platforms that could support it, including CF.

So the only real issue I see here is the one of the default value for --workers, which is currently equal to the number of processors in the system + 1.  That was an arbitrary choice, but seemed, from our testing, to be reasonable for the vast majority of test scenarios.  Since this value could easily be changed from the command line, the default really did not seem to be a significant issue.

**Comment by oznetmaster** [comment]
As to the idea of a way to obtain the current NUnit version, this seems something for a separate issue where it can be discussed by itself?

If you would like to open one, I am sure it would elicit a lot of useful comments.

**Comment by pavelsher** [comment]
Well, I personally, not an expert in .NET, I work mostly with Java code. And in Java we also have frameworks which can run tests in parallel. Many times we tried to run tests in parallel in Java projects, and reasoning was like yours - tests spend lots of time on I/O operations, why not run them in parallel?

Unfortunately more often than not, this caused serious instabilities, mostly because tests which look like independent in fact depend on some common data placed in static structures which are shared among all of the threads. This applies not only to our code, but to some libraries as well, hence it's not easy to fix. The only option for parallelization that we're considering for our project (TeamCity itself) is to run tests in different processes instead of different threads. But even with processes some very subtle issues can occur for instance when tests create temporary files, or operate in the shared directory, or obtain connections to the same database schema and then change it, etc, etc.

So I'd not be so sure, that this default is safe, our experience shows that it's not. And there is a reason why parallelization is not enabled by default in popular testing frameworks in Java (at least to my knowledge).

Again, parallelization per se is useful, I am speaking about default only.

**Comment by oznetmaster** [comment]
Why don't you open a separate issue about changing the default number of workers to 0 (disabling parallelization)?

Then an open discussion can be had on that single topic.

**Comment by CharliePoole** [comment]
@NikolayPianikov There are so many things already in this "issue" after only eight hours that I hardly know how to start responding! I think what I will do is add separate comments that address different things that have been raised. This can be a "mother story" and we can spawn from it some individual issues that we can actually work on, since I don't think there is really anything for us to do in a single merge that will deal with the entire set of questions you raise.

@pavelsher @oznetmaster I haven't really even read your comments yet, so please be patient. I will get back to you. :-)

In this comment, I am going to address the general question of backward compatibility and how we have dealt with it. First, I think you need to have some history to understand...

I first outlined the plan for NUnit 3.0 in 2007. It was presented as a design document and a set of ideas at the Mono Gathering in Madrid that year. I called it NUnit-Xtp (for Extended testing platform) and presented it as a new program, not an upgrade of NUnit. I explained then that I would keep using the name NUnit because the new program would be _mostly_ compatible and for sentimental reasons (as well as marketing).

That was a long time ago, but I've been presenting it that way to everyone for a long time. You can find many references where I've told people on our mail list that NUnit 3.0 is actually **not** an upgrade of NUnit 2.x. That's why we continued for all these years to put out new releases of the old NUnit. I count 22 releases with many major improvements since I first started working on NUnit 3.0. That's one major reason why NUnit 3.0 has taken so long to come out.

[By the way, I now regret I didn't just drop NUnit 2 and work on NUnit 3, but that's another story.]

NUnit has an exemplary history of backward compatibility. I know that's true because people have criticized me over the years for maintaining _too much_ backward compatibility. The current stable release of NUnit, NUnit 2.6.4, is still capable of running tests under .NET 1.1 and (with some tricks) even .NET 1.0. It can run tests that were compiled against NUnit 2.0 going 15 years ago. So we do have some understanding of and sympathy for users' need for compatibility.

So, we decided to break compatibility in NUnit 3.0 only after a lot of time and thought and, as I said, we announced we were going to do it seven years ago. People who were listening at the time or who did some research should know this because we have published all the info about it.

That's not true with TeamCity. We do not know what you are doing unless you tell us. We can't add features needed only by you unless you ask us. Specifically, I did not know you guys were already supporting the beta 3.0 until a short while ago when you mentioned it in a post. I did not know that your interface was through the console command line until you complained about changes to the command line. Honestly, I didn't know you worked for JetBrains until after I saw that your merge requests were all about team city and wrote to ask you! Naturally, having worked in industry, I know that proprietary companies don't like to tip their hand to outsiders, but I think you have to accept that there are some consequences of that silence.

Many of the decisions you are now unhappy with (which I'll get to in detail in separate comments) were discussed at some length before we made them. They were discussed openly on our mail lists and in github issues. To my knowledge, you guys have not participated in these discussions. It seems to me that it could have been beneficial to you if you had taken part. You were and are welcome to and I hope you will in the future.

So finally, how do we view NUnit 3.0 compatibility? Just as you said, I define it in terms of "APIs." A long time ago, I enumerated five of them we would have to deal with and set some goals for them:

1. The interface used by programmers writing tests, that is the Asserts, Attributes, Constraints and their semantics. This is the key api for NUnit and the one we worry about the most. We have tried to keep this within 90-95% compatible and I think we have succeeded. However, we have allowed a number of breaking changes, which are documented, **even though this is our most important "api"**

2. The extensibility interface, primarily related to NUnit 2.x addins. We have completely thrown this out, breaking compatibility for those who extend NUnit.

3. The interfaces used by folks who want to run tests programmatically. This was an easy one. It never existed before so there is nothing to break. :-) NUnit 3.0 creates two new apis - one at the framework level and one through the engine - for running tests. Of course, we will have to worry about compatibility of these as we move ahead with new releases.

4. The command-line interface for our executables, like nunit-console. Frankly, we have treated this as less important because the commands are generally typed in by users - either at the command-line or into a batch file. They will have to learn new commands and edit their scripts. The options we have removed are all documented and include a lot of obsolete stuff we have been waiting to get rid of for a long time - waiting, in fact, for a major release.

5. The appearance of the NUnit Gui, what functions are available, etc. I consider this the least important in terms of compatibility so long as equivalent functionality is available.

As you see - and as you have noticed - command-line compatibility is pretty low in our list of priorities. Frankly, I had not give much consideration to the fact that folks were writing programs that generated the command-line. An earlier heads-up might have changed some things that you don't like, but probably not all of them.

It seems to me that Jetbrains also made some decisions about compatibility. From what you write, I guess you decided that the users should be shielded from any consequences of this change. That sounds good until you realize that we are telling users that the change is _not_ a simple upgrade. You are using and depending on NUnit, but are trying to make it work in a way that we are not. I think it's too bad that there was no discussion of this up front.

If there had been a discussion, I would have suggested that you simply ask the users to tell you, in their templates, whether they were using NUnit 2 or the new, non-compatible NUnit 3. I might be wrong, but it seems as if this would solve all your problems without any serious inconvenience to the user. After all, they have to tell you if they are using NUnit in the first place!

However, my goal isn't to tell you how I think you should have done it and you probably wouldn't want to know anyway. It's about how we can move forward. I'll try to do that in further, more specific, comments about the problems you raise.

**Comment by CharliePoole** [comment]
@oznetmaster Regarding --workers=1 versus --workers=0 - I agree with you as the code stands now. I have wondered about whether we should keep --workers=0 in the future, however.

One advantage of --workers=1 is that you get a separate queue and thread for tests that must run in the STA. OTOH, it may be an advantage of --workers=0 that each test requiring the STA is run on its own separate thread. All in all, it's a minor detail that they could probably change easily enough.

**Comment by CharliePoole** [comment]
Let me address the parallelization issue.

When we first introduced parallelization (2.9.6, November, 2013) there was extensive discussion about what the default should be. A lot of people put in their views and we made a decision. We have been running code that implements that decision for two years now. It's a bit frustrating to have it called into question at such a late date.

Nevertheless, here's what we decided. Note this is only for those platforms on which we support parallel execution, which of course includes the platforms TeamCity runs on.

The default maximum number of threads is Max(ProcessorCount, 2). This causes some overhead for those test assemblies that don't use parallelism at all, but it's pretty small, amounting to the creation and destruction of a few threads that are never used. In a future release, we will probably wait till the threads are actually needed to create them.

The default for ParallelScope on tests is None, meaning that no tests run in parallel. A user has to put an attribute on a test or on the assembly before any tests are run in parallel.

If you think this should be changed, please explain how. If you have seen NUnit operating in any way that contradicts the above, that's what we call a "bug" :-) Please file an issue in that case.

**Comment by CharliePoole** [comment]
@pavelsher Just a small comment on one thing you wrote. 

"We took the risks to integrate with the beta version..."

I hope you noticed in my earlier long post that I and the other NUnit guys didn't have any idea that you were doing that. I'm not saying you should have told us, only that it might have been a way to reduce the amount of risk if we had known and had known what you were relying on.

Just a thought for the next time!

**Comment by CharliePoole** [comment]
Regarding NUnit version... Somebody please file a bug report asking for a version option... at least if you want it!!! It's trivial both for us to agree to this and to implement it.

**Comment by CharliePoole** [comment]
Regarding "include" and "exclude" options on console program.

We were faced with a difficult choice here. We have been planning to get rid of these for a long time but didn't yet have the code. If we release 3.0 with those options, then we are forced to continue maintaining them until we decide on another breaking release, which might be ten years.

We are following a philosophy that says we maintain backward compatibility over a long time. The other side of the coin is that we don't want to put anything in 3.0 if we are not willing to maintain it for a long time. This has led to some last-minute changes that I would have preferred to make in an alpha stage because, to me, beta means features should not normally be removed, only added. Nevertheless, we made the decision. The where clause will give users a much friendlier way to select tests and eliminates the problem our former "category expressions" had when the category name contained a hyphen.

I continue to wonder if this might not be resolved most simply if you treated NUnit 2 and NUnit 3 as completely different frameworks, with different options available to users. (I know that's not my business but it seems so simple!)

**Comment by CharliePoole** [comment]
@NikolayPianikov This is a repetition of an earlier comment, but it's important.

You say "Parallelization is on by default..." This is not what is supposed to happen. If you check the docs, you'll see that nothing should run in parallel without an attribute telling us to do it. Please check what you are seeing and file a bug if it's not working right!

**Comment by CharliePoole** [comment]
@NikolayPianikov Regarding the program name... nunit3-console

I find myself not liking this much as a user. I suppose I'll get used to it though.

I think I am regretting the change as well as it's timing. Strangely, we imagined it making life easier for CI servers!!! We thought that you would want to decide which of the two to run and have a different name for each. I suppose we were guessing what you needed.

@nunit/core-team What do you think about this?

**Comment by CharliePoole** [comment]
@NikolayPianikov NUnit4-console? Will you still be around when that time comes?

NUnit31-console, NUnit32-console? Now you're being silly. Besides, that would be a breaking change, so we would be forced to call it 4.0 by our own rules.

I'm pulling your leg, of course, but the point is that this is the first major break for NUnit since the year 2000. If this makes you think we will do it again in a month or a year, then you are not paying attention to the evidence.

**Comment by rprouse** [comment]
The rename was actually spawned from the Cake build tool so that the NUnit
task would not mistakenly try to run NUnit 3 with the old, incompatible
command line if it was found in the path. The thinking is that NUnit 3 is a
breaking new release, so it requires a new task runner within Cake or other
build tools. Maybe a version command line option would help with that, but
it would require running the app to verify.

Like @charliepoole, I see NUnit 3 as a new product that requires new
runners. That is why the NUnit team released separate Visual Studio
adapters and other 3rd party runners like TestDriven.NET are adding new
tasks or runners to support nunit 3. I would think that doing so in
TeamCity would also be the best approach because it would allow users to
have options that mirror the command line options that they will become
used to. In my opinion, trying to map NUnit 2 options to NUnit 3 options
will be more work than is necessary and will only end up confusing users.

On 2 November 2015 at 16:21, CharliePoole <notifications@github.com> wrote:

> @NikolayPianikov <https://github.com/NikolayPianikov> Regarding the
> program name... nunit3-console
>
> I find myself not liking this much as a user. I suppose I'll get used to
> it though.
>
> I think I am regretting the change as well as it's timing. Strangely, we
> imagined it making life easier for CI servers!!! We thought that you would
> want to decide which of the two to run and have a different name for each.
> I suppose we were guessing what you needed.
>
> @nunit/core-team <https://github.com/orgs/nunit/teams/core-team> What do
> you think about this?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nunit/nunit/issues/976#issuecomment-153160332>.
>



-- 

Rob Prouse



I welcome VSRE emails. Learn more at http://vsre.info/

**Comment by CharliePoole** [comment]
Summary of the specific issues:

1. Getting the version. Easy enough. File a request.

2. Name of the console program. I'm waiting for team feedback. If we want to reverse it, we will have to do so quickly.

3. Loss of include/exclude. I'd really like to keep them gone. They have intrinsic problems that we have been aware of for a long time. And anyway, the content of the options would be different for NUnit 2.x and 3.x because 2.x has category expressions while 3.x only supported comma-separated categories through beta 4. I **really** don't want to support both category expressions and test selection language.

I did this summary so you guys could tell me if I missed any specific point you want considered.

**Comment by CharliePoole** [comment]
@rprouse I'd like our main focus to be on users though. And personally, I am finding it awkward to type a new command that's one letter longer. Silly I guess.

Of course, you'll say that there are few users working at the command-line these days, but those are my guys! I'm strongly leaning to changing it back. That will mean an RC2 but I think we will have that anyway if we follow the rule that there are no code changes between rc and final.

**Comment by oznetmaster** [comment]
I personally think this is a silly issue.

Do people really type these out on the command line anyway, and not use scripts?

However, if it causing such an upset, then change it back.

**Comment by oznetmaster** [comment]
I also indicated that a request should be filed for a "version api".

I think that now that --include and --exclude are gone, they should stay gone.  It took me less then an hour to change every program that generated nunit command lines (like my remote gui) to use the --where option (and that was five different programs).  This is not rocket science.

**Comment by rprouse** [comment]
@charliepoole, you could always give up nunit3.exe from the gui and make it
really easy to type :)

Personally, I've always found the - in the name just as annoying as the new
3. At least nunit.console.exe would be consistent with our other assemblies
and the . is only one row from the home row for touch typers :)

On 2 November 2015 at 16:55, Neil Colvin <notifications@github.com> wrote:

> I also indicated that a request should be filed for a "version api".
>
> I think that now that --include and --exclude are gone, they should stay
> gone. It took me less then an hour to change every program that generated
> nunit command lines (like my remote gui) to use the --where option (and
> that was five different programs). This is not rocket science.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nunit/nunit/issues/976#issuecomment-153169500>.
>



-- 

Rob Prouse



I welcome VSRE emails. Learn more at http://vsre.info/

**Comment by CharliePoole** [comment]
@oznetmaster Well I type the commands, which influences my thinking. I guess Rob does as well.

@rprouse Actually, we have consistently used a hyphen in exes for many years. That's why it's nunit-agent, for example. We've always been outside of the "standard" for Microsoft executables, which typically use dots. I think our users are used to it by now.

I don't think the problem is so much the row the keys are on as the fact that one has been typing nunit-console for the past 15 years and now has to change. Other people have probably not been doing it so long.

If we used nunit for the console, what would we use for the gui? We got rid of nunit-gui.exe many years ago.

**Comment by NikolayPianikov** [comment]
I've created #978 about version

**Comment by pavelsher** [comment]
@CharliePoole you made several valid points, let me address them one by one.

Why we did not communicate properly about NUnit 3 support in TeamCity: well we did not have much problems with this integration, at the time when we integrated with beta 3 nothing told us that there will be major changes in command line after that. We did not expect it from Beta version, but this is already discussed, so we'll be smarter next time :)

Speaking about NUnit as a new tool, I personally not sure this is a widely known fact, because, for instance, we who maintain integration with NUnit during the last several years were not aware about this aspect until this year. I think significant percent of people understands that next version of NUnit is more like a new product, but others, those who do not watch closely for the evolution of the tool, probably not.

Why we prefer to integrate with command line? Because there is no reliable way to run NUnit with our own listener. Since our listener has to be strongly signed we end up with necessity to provide a listener for every bugfix version of NUnit which is not maintainable. Not only that, but we also have to patch MSBuild scripts when they are started by TeamCity to pass this listener to NUnit task which is also non trivial and fragile. Well, we could compile our listener before starting a build, but this is just too complicated to work reliably. So this is why we prefer to use command line and since NUni3 now is able to output service messages in console when it detects presence of TeamCity agent, this makes our integration much simpler. Finally, we prefer to start tests in a way which is repeatable by the user, so that if something fails, user could execute exactly the same command and see that TeamCity agent does not affect execution of tests. I hope it describes why CLI is important for us.

BTW, thank you for accepting our patches for service messages, they made our life a lot easier!

Loss of includes / excludes - currently we workarounded this problem by dropping compatibility with NUnit 3 beta 3, so we're now using --where flag for all NUnit3 versions. 

Change of the tool name is still a bit painful for us, because we're not sure what will be the name of the tool in the future. So now we decided to ask users to provide path to .exe itself, this way if NUnit team decides to change name of the .exe file one more time, our integration will continue working if user changes path to the tool and the rest of CLI is the same.

To sum up, the important outcome of this is that for CI tools CLI is important. Note, that I am not speaking about TeamCity only. Other CI servers use CLI as well because other ways are way too complicated.

**Comment by oznetmaster** [comment]
I use the command line exclusively for my remote execution GUI which runs NUnit 3.0 on CF devices remotely (over tCP/IP)

I have not found any of the changes made from version to version to be a problem.  In most cases, I can either determine the version from the command line and its responses, or if not,  add a menu item to select different framework versions.

Adding support for a new version has never taken me more than about 30 minutes for both the programming and testing.  As an example, changing from --include and --exclude to --where took about 15 minutes (three lines of code).

**Comment by CharliePoole** [comment]
@pavelsher Of course, if we had understood just how you were using NUnit, that could have been one added factor to discuss when we made some changes. For example, the removal of --include and --exclude was debated, with various pros and cons. Nobody ever said "This will impact TeamCity, which generates a comand-line" and so that was never weighed among the choices. But as you said, we all have to do better next time.

About CLI integration: yes, it makes perfect sense. The APIs for running NUnit programmatically are intended for test runners not CI servers. That said, folks who use the engine api do not have to recompile for each new release. The api assembly, which is the only one you actually reference, is signed and is at version 3.0.0.0 permanently. This is part of the design. Of course, if we were actually not able to keep it at the same version, then a re-compile would be needed, but I think we will be able to do that for two reasons:
1. Most engine changes will be accompanied by changes to the value of strings (usually xml) passed to the API.
2. If no features call for it, we will only modify the API additively.
However, this is more for your colleagues who work on R#.

Regarding awareness of NUnit 3 as a new tool... Yes, this is definitely a problem. NUnit has a very large body of users and most of them don't keep up with the tool's progress. Many people will - unfortunately - be surprised when the time comes to upgrade. We gave that a lot of thought but we felt we had no other choice if we want NUnit to move forward without carrying all the weight of previious years. 

That's why we have not encouraged people to upgrade thoughtlessly. Our nuget packages, in addition to being marked as pre-release right now, have entirely different ids from their predecessors. I'm writing release notes now for the final release, which will emphasize that this is a new program and explain all the differences from NUnit 2.6.4 rather than just giving the last changes from the beta. This is why I was surprised that you want to make the change transparent to the users. This does not seem possible to me, because of the large differences. Most of the differences are not visible to you, of course, but they are to the users who have to modify their tests.

As you have seen, we are discussing whether to revert the change of name for nunit-console. As a user myself, I'm not so happy with it now that I am using the RC. On the other hand, I'm not a typical user, since I work so much at the command line.

**Comment by NikolayPianikov** [comment]
Maybe it has already been said, but

include/exclude: I missed discussion about include/exclude instead "--where". I saw some discussion about --where but did not see "--where" is going to replace include/exclude. We solved this issue by converting 2 fields to one. But it works differently for different version of NUnit 3 beta. Now we are getting a version from NUnit-console's std-output before next steps. Different names makes the using of NUnit-console more difficult.

parallelizable tests: Indeed the test could be run in parallel mode when it marked by [Parallelizable] attribute, but to run it in parallel, the user must explicitly allow this by an argument "--workers>1".  This argument (--workers) has another value by default now. This argument could be applied for assembly too as I know. Let's hope that people who use that attribute, knew what they were doing. So we can close this issue.

**Comment by CharliePoole** [comment]
@NikolayPianikov 
include/exclude: It may be that some things were not being fully explained in the discussion. Those of us working on the code already knew that the 3.0 include/exclude feature was a simplified, temporary port of include/exclude from 2.0. We put it in so we could run tests while working on the new approach to filtering tests. Sorry! We should try to discuss everything in a way that people seeing the discussions can get the full picture.

parallelizable tests: Yes, that's exactly how we thought about it. It is not possible to use the command-line to force tests to run in parallel because they may not run correctly. OTOH, it is possible to force them to NOT run in parallel in case of a problem. As for our hopes, it's always an issue with advanced options that people could use them without fully understanding. We hope they won't and we'll keep working on the docs to make it clear how to use parallelization.
